doc/design.md

```diff
@@ -0,0 +1,60 @@
+
+## Requirements
+
+### Functional requirements
+
+* Provides information on virtual machines running on the host.
+* Runs on host if the host supports the KVM user-space API and the underlying hardware has capabilities to support virtualization.
+  * For compatibility with edge devices without virtualization support, `vmd` server may run to inform client about lack of capabilities.
+* Per virtual machine information includes at least: memory and CPUs. Other resource information is optional - if in use.
+* Controls for virtual machines: shutdown (if running from boot), start-up (if not running from boot) and resource reallocation (changes to CPU/memory and possible other resources - if in use).
+
+### Non-functional requirements
+
+* Server performance overhead should be optimized. Information is served upon request to avoid polling overhead.
+* Implementation with secure systems programming: Rust for memory safety.
+* Clients are authenticated (e.g. mutual TLS) to secure highly privileged information and operation on secure host.
+* Communication between `vmd` and clients is encrypted (TLS).
+
+## Design principle
+
+### Loose coupling
+
+`vmd` as a system component is weakly associated with both client applications and underlying interfaces. In practice, this could be achieved for example with RESt API where get/set operations are atomic per operation. Other loosely coupled interfaces than RESt API can be considered as this design is at proposed state.
+
+Other area of loose coupling is the system graphics architecture. `vmd` does not provide clients with information on the guest virtual machine graphics or GUI views. In practice, the host of `vmd` may or may not have Wayland compositor *but* the `vmd` does not provide its' clients any information on the Wayland or X clients. This decoupling is:
+* practical - a ghaf system may be headless
+* performance-driven - routing Wayland protocol through `vmd` is overhead and
+* security - `vmd` does not need to "see" what's in the guest virtual machines Wayland client views
+
+If `vmd` clients need to provide views to either Wayland client views or the system view of Wayland compositor, the `vmd` clients are encouraged to obtain the information from those sources directly.
+
+### Encapsulation
+
+`vmd` encapsulates host internal sources of information and provides them to the clients over networked interface (TCP/IP). Encapsulation means that the host interfaces, such as KVM API or systemd APIs or rust-vmm crates, are not directly exposed to the `vmd` clients. This mechanism allows `vmd` server internal implementation to change without changes to external interface.
+
+### Distributed
+
+`vmd` backend supports distributed clients. Clients may run within the same system (inside the host or in guest virtual machines) or may access the `vmd` using secure connection over the internet. This supports graphical, command-line and web app clients - depending on the need and use case.
```

Comment:
Should be there more details for every client app type (GUI, CLI and Web)?

Reply:
I do not agree on that as this is server side design proposal. If we agree on this high level design - REST API - we can start designing and prototyping the REST API. The REST API would be the same for every client app type and the server does not consider the presentation layer (GUI) of the backend data. That would be part of the client design.

Even the host local CLI test application (over SSH in the diagram) would use the same REST API. I would be inclined to say that non-SSL would not be supported even for the local CLI as it would pose a security threat to the system. In local CLI testing we could use test SSL certificates with REST API - like the remote client would use. Just shared test certs to make testing easier.